[IMAGING-355] Large animated GIF takes too much heap memory in getMetadata

andrewmkhoury opened the pull request:
IMAGING-355 - Large animated GIF takes too much heap memory in getMetadata

kinow commented on src/main/java/org/apache/commons/imaging/formats/gif/GifImageParser.java:
```diff
@@ -407,7 +407,7 @@ public byte[] getIccProfileBytes(final ByteSource byteSource, final GifImagingPa
     public ImageInfo getImageInfo(final ByteSource byteSource, final GifImagingParameters params)
             throws ImagingException, IOException {
-        final GifImageContents blocks = readFile(byteSource, false);
+        final GifImageContents blocks = readFile(byteSource, true);
```
Maybe this `true`/`false` (`stopBeforeImageData`) should instead be retrieved from the `GitImagingParameters` object?

garydgregory replied:
That's a good point!

I wonder if this data is hard-coded to be skipped it will be unavailable for other API calls once the object is constructed. That would be bad. But if it can be paramerized, then the client call.site can decide.

garydgregory replied:
**Note**: All current callers of `org.apache.commons.imaging.formats.gif.GifImageParser.readFile(ByteSource, boolean)` pass false for `stopBeforeImageData`.

Would it make for all call sites to pick up this value from `GifImagingParameters`? Or, would this cause some use-cases to break. We will need to make sure to unit test whatever code we change or add.

andrewmkhoury replied:
`stopReadingBeforeImageData` or `skipImageData` works.

andrewmkhoury commented:
@garydgregory and @kinow I implemented the changes per your request.  However, I am not sure how much value passing stopReadinngBeforeImageData adds.  getImageInfo already has a unit test and I added a test for getMetadata.
* Add stopReadingBeforeImageData to GifImageParameters
* Add unit test for IMAGING-355

garydgregory commented on src/main/java/org/apache/commons/imaging/formats/gif/GifImageParser.java:
```diff
@@ -407,7 +407,11 @@ public byte[] getIccProfileBytes(final ByteSource byteSource, final GifImagingPa
     public ImageInfo getImageInfo(final ByteSource byteSource, final GifImagingParameters params)
             throws ImagingException, IOException {
-        final GifImageContents blocks = readFile(byteSource, false);
+        boolean stopReadingBeforeImageData = true;
```
You've changed the default behavior here which makes the parameter a forced usage, which might be a surprised.
